### PR TITLE
feat(planning): planning does real planning — one pass, host-verified prerequisites (#337)

### LINKED ISSUE #337: spine: a task advances itself through the board
Part of the spine epic **#183**, §4. **UNBLOCKED as of 2026-08-05** — #242 shipped in three PRs (#342 the port and storage, #357 the write path, the read surface), so an attempt is now a durable thing with a status, a trace and a cost. #333 (approvals linked to their card) and #334 (the Done column was unreachable) have also closed since this was written.

With those in, this is the main remaining gap between "a person hand-carries every card" and the spine running itself.

## What is missing

A card moves because a person drags it. The one automatic edge in the system is `task_enters_in_progress` — entering In Progress dispatches a cycle. Nothing else moves on its own: a finished run does not advance its card, a blocked run does not park it, and nothing ever enters Planning.

So the spine's promise — type one sentence, watch it walk to a deliverable — currently requires an operator to hand-carry the card through every station.

## What this should do

Close the loop between a run's status and its card's column.

```
Todo ──plan──▶ Planning ──dispatch──▶ In Progress ──┬──▶ In Review   (a person must act)
                                                    ├──▶ Paused      (anything else)
                                                    ├──▶ Done        (finished, output attached)
                                                    └──▶ Todo        (failed or unplannable, reason on the card)
```

The mapping is not invented here — **#183** settles it and **#242** encodes it in `RunStatus`:

| Run status | Card lands in |
|---|---|
| `waiting_approval` | In Review |
| `paused` | Paused |
| `succeeded` | Done |
| `failed` | Todo, with `run.error` written to the card's note |
| `cancelled` | Todo, with the cancellation reason on the card |

The `failed` row is the settled decision from #183: **a task that cannot be planned goes back to Todo with the reason on the card**, not into a stuck column. `RunRecord.error` is what carries the reason — that field exists precisely so the card can say why.

---

## Planning — SETTLED (2026-08-04)

This was the open question. It is now decided: **Planning does real planning.** It is not a status-only station a card passes through.

### What the planning pass produces

Everything a person would want before work starts, written onto the card:

- **Decomposition** — the request broken into the steps that actually have to happen, in order.
- **Estimates** — what each step is likely to cost and how long it should take.
- **A detailed description on the doc** — the card stops being one sentence and becomes a brief someone could hand to a teammate. This is the main visible output.
- **Prerequisites gathered** — which connections, credentials, files, inputs and permissions the task needs, checked against what the company actually has, so a missing one surfaces *before* an agent burns a turn discovering it.

### Shape it after our own planner

The model to copy is the planner we already use for local development work: gather evidence first, then emit a structured plan — the problem restated, the approach as ordered steps, what it will touch, the risks and trade-offs taken knowingly, how it will be verified, and an honest scope call. That structure exists because it is what makes a plan reviewable rather than a wish. The same applies here.

The important property to carry over is **evidence before prescription**. A planner that starts prescribing without first reading what is actually there produces plans that drift toward whatever it already assumed.

### No run record, no run lock

**The planning pass does not get its own `RunRecord` and does not take the per-company cycle lock.** Two consequences, both intended:

- Planning does **not** serialise behind in-flight cycles. Several cards can be planned while another card's agent is working. Planning is not a dispatch and should not queue like one.
- There is no attempt row for a plan. The plan's output lives on the card, which is where it is useful; #242's run records stay reserved for actual work attempts, so an operator reading the Attempts list sees turns that did something, not bookkeeping.

**Two things this makes the implementer's job:** the planning pass still calls a model, so its cost needs attribution through the usage path even without a run row; and a planning failure has nowhere to be recorded as a run status, so it must land on the card as the reason — the same channel as the `failed` row above. Both are settled below.

### Cost attribution — SETTLED

**Charge planning to the company bucket, under its own sample kind. Not to the assignee.**

`UsageSample` carries an `agent` and a `kind`, and `UNATTRIBUTED_AGENT` (`"company"`) already exists for exactly this case — its own doc says *"the tokens were genuinely spent, so the totals must include them even when the breakdown cannot attribute them."*

Charging the assignee is wrong twice over. **Planning is often what picks the assignee** — decomposing and gathering prerequisites is how you learn who should do the work, so at plan time there frequently isn't one. And since #304, per-agent daily caps are *enforced*: planning charged to the assignee would eat that teammate's cap, so a teammate near its limit could make its own card unplannable, and the operator would see a spend refusal for work that never started.

It needs a **new `SampleKind`** rather than reusing one. The codebase already made this argument for `SearchCall`: reusing a kind whose semantics differ both drops the real cost and invents a false signal in the derived chart.

### Bounding the spend — SETTLED

**One planning pass per entry into Planning. No automatic retry.** A failed plan returns the card to Todo; re-planning requires the operator to move it back.

`"company"` is not a roster agent, so #304's per-agent cap does not reach it — planning would otherwise be unbounded. Bounding it by construction caps volume at cards the operator actually creates, and avoids adding a second budget knob to configure and get wrong.

Note that #343 has since made caps operator-settable from the Console, so a dedicated planning cap is no longer the dead-end it would have been. It is still the second choice: bound by construction first, add a knob only if that proves insufficient.

---

## Why it blocks on #242

Two reasons, both concrete:

1. **The column mapping reads a run status.** Without a durable per-attempt status there is nothing to map from — today the only in-flight signal is a volatile in-memory registry that dies with the process.
2. **A crash must not strand a card.** #242's boot reaper settles an attempt that died with its process, so its card can leave In Progress. Without it, a restart leaves cards parked mid-board with nothing to move them.

## Acceptance

- A run reaching each terminal status moves its card to the column above, without an operator.
- A `failed` or `cancelled` run's card lands in Todo with the reason visible on the card.
- A card in Paused or In Review is never moved automatically — only the person, or the unblocking event, moves it.
- Killing the host mid-run and restarting leaves the card in a truthful column, not stranded in In Progress.
- Re-dispatching a card that came back to Todo starts a **new** attempt, not a resumed one.
- A task may enter In Review more than once across attempts (#183).
- **Planning produces a decomposition, estimates, a detailed description and a prerequisite check on the card** — a card that has passed Planning reads as a brief, not a sentence.
- **Planning creates no run record and does not block on the cycle lock** — two cards can be planned while a third is being worked.
- A planning pass that cannot proceed returns the card to Todo with what it needs, so the operator can revise the prompt.

## Related

- #183 — the spine epic (§4)
- #242 — run records **(blocker)**
- #301 — shipped the Planning column inert; this makes it live
- #243 — single-use, argument-exact grants, which is why In Review is re-enterable

### PR DESCRIPTION
**Pull request by oxoxDev**
## Summary

The second half of #337: the **Planning station**. PR #408 shipped the
status→column half (`column_for_settled_run`, the guarded `advance_settled_card`,
the three stranding paths). This makes the Planning column actually do something.

A card entering `planning` edge-fires exactly one pass: a **single tool-less
model call** that writes a structured brief onto the card and then settles the
card itself, three ways.

| Outcome | Landing | Card carries |
|---|---|---|
| planned, nothing blocking, valid assignee | `in_progress` (the dispatch edge fires) | the plan |
| planned, a hard prerequisite missing | `todo` | the plan **and** the named gap |
| the pass failed (model error, timeout, unparseable output) | `todo` | the reason only — no plan |

Full contract: `docs/spec/runtime/planning.md`.

## API Or Behavior Changes

**Behaviour — Planning is now a spend gate.** Before this, `in_progress` was the
only column whose entry cost money. There are two now. The drag into Planning is
opt-in per card, and on success the card hands straight on to `in_progress`
without asking again — so the drag is informed consent to both. `todo →
in_progress` still dispatches unplanned, so nobody is routed through planning.

**Wire — additive only.**
- `TaskRecord.plan: Option<TaskPlan>`, serde-default + skip-if-none, the
  `origin_chat_id` / `parent_task_id` precedent. No stored board needs migrating
  on fs, sqlite or mongodb.
- The REST card DTO projects `plan` verbatim. **Intake cannot set it** — the
  create body has no field for one, so a client cannot post a card that already
  claims to be planned or forge the verdicts that decide whether it dispatches.
- **GraphQL deliberately omits `plan` in v1.** The console reads the board over
  REST; the SDL snapshot pins the omission, so adding it later is a deliberate
  edit rather than a drift.

**Usage — new `SampleKind::PlanningCall`.** Backend-internal (the kind never
crossed the wire), but it *is* counted by `metering::tokens_in`, so planning
tokens now consume the #108 capability-tier ceiling. Excluding them would let a
company plan indefinitely after crossing the budget meant to stop it. This is a
correction to the approved plan, which asserted the tier already counted it —
`tokens_in` filtered on `Inference` only, so I made the code match the decision.

### Evidence before prescription — by inverting who gathers

Giving a planning agent tools to go and look is a tool loop, which is an agent,
which is the dispatch planning is supposed to precede. So the **host** gathers
the evidence deterministically and the model only synthesises: the roster with
each teammate's effective grants, the connections, the MCP union, a bounded
workspace listing, the skills, the approval policy. `ModelRequest::tools` is
empty — there is no loop and no path by which a pass can act on the world.

The pass does **not** use `HarnessPool::run` (per-agent cap gate, agent cost
attribution, tool loop) nor the cycle/`Brain` (per-company serial lock). One
direct `ChatModel` call under a hard 120s deadline.

### The model claims, the host verifies

The model emits `{kind, name, why}` and **cannot** emit a status — `PrereqClaim`
has no such field, so the asymmetry is enforced by the type, not by the prompt
asking nicely. The host stamps every verdict:

| Kind | Checked against | `missing` reads as |
|---|---|---|
| `connection` | the same projection `GET …/connections` builds | "slack is not connected — connect it from the Connections tab" |
| `composio` | that projection's `via`, plus token presence | distinguishes *no credential* from *no account* |
| `mcp` | manifest `[[mcp_server]]` ∪ the runtime index — **both** halves | in neither; a *disabled* server is its own verdict |
| `credential` | presence only — the mail handle, or a key that exists | "no outbound email is configured" |
| `file` | the workspace tree, trailing-segment match | "references a path that is not in the workspace" |
| `permission` | the **working** teammate's manifest grants + `[policy]` | not granted, or granted under read-only |
| `assignee` | the roster resolver | nobody by that name is on the roster |
| anything else | nothing — and the host says so | — (always `unknown`) |

**Verdict taxonomy:** only `missing` blocks. `needsApproval` is a warning (the
operator will be asked at the moment of use, which is the design) and `unknown`
means the host could not check.

**An inventory that could not be read yields `unknown`, never `missing`** — a
Composio outage must not make every card in the company unplannable.

Permissions read the **manifest only** (allow-list, agent tools, `[policy]`) —
never `runtime/grants.rs` or `harness/policy.rs`, which are per-call runtime
machinery that legitimately moves between planning and dispatch. A desk resolves
to its **lead**, who is who actually runs the turn.

**Only names and booleans enter the prompt.** No credential value is read, let
alone rendered.

### Concurrency: three layers, no lock

A pass mints no run row (there is no agent turn, no trace, nothing to steer — a
row would be a phantom attempt on the card's timeline) and takes no cycle lock
(that lock is held for a whole agent turn; taking it would park the company
behind a planning pass). Three cheaper guarantees replace them:

1. **Edge-firing at the single write site.** The trigger is the *transition*
   inside `upsert_task`, which every REST task mutation funnels through. A card
   re-saved in Planning — an edit, or the pass's own note append — is not a
   transition. This gives *one pass per entry, no retry* by construction.
2. **A per-company in-flight set.** Covers what the edge cannot: dragging a card
   out and back in *is* a second genuine transition. Drop-guard release, so a
   panic or an early return cannot leak a claim.
3. **An optimistic settle guard.** The pass captures `updated_at_millis` before
   the call and requires it unchanged *and* the column still `planning` at
   settle. Any operator action bumps the stamp, so the operator's move wins and
   the whole pass is discarded. The tokens stay metered — they were spent.

**Crash recovery**: no run row means no orphan for the boot reaper, so
`sweep_stranded_planning` reads the board directly at boot and returns Planning
cards to To-do. Sound because Planning is transient by construction; suppressed
on a rebuild, like the two sweeps beside it.

### Cost — the company pays, not the assignee

`SampleKind::PlanningCall`, `agent: "company"` (`UNATTRIBUTED_AGENT`),
`run_id: None`; money posts through the existing `inference.spend` entry.
Charging the assignee is wrong twice: planning is frequently what *picks* the
assignee, and since #304 a teammate near its daily cap would make its own cards
unplannable. `"company"` is not a roster agent so it is uncapped by #304, but it
does count toward the #108 tier ceiling. `bucket_usage` needed no change.

## Submission Checklist

- [x] `cargo fmt --all -- --check`
- [x] `cargo clippy --locked --no-deps --all-targets -- -D warnings` (default lane — what CI runs)
- [x] `cargo clippy --locked --no-deps --features openhuman,tinycortex --all-targets -- -D warnings`
- [x] `cargo test --locked --lib` — 1445 passed
- [x] `cargo test --locked --features openhuman,tinycortex --lib` — 2099 passed
- [x] `cargo check --locked --all-features --all-targets`
- [x] `cargo test --locked --features sqlite --lib store::sqlite::test::conformance` — 17 passed
- [x] `git diff --stat Cargo.lock` empty
- [x] Frontend: `npm ci`, `npm run typecheck`, `npm run typecheck:e2e`, `npm run build`
- [ ] `cargo build --all-targets` — N/A: superseded by the two `--all-targets` clippy lanes and the all-features check above, which compile strictly more.

### Verification on a real binary, both surfaces

`cargo build --features openhuman,mcp,telegram` + `npm run build`, served against
`companies/e2e_harness`. No live inference credential exists in this
environment, so the provider is a **local OpenAI-shaped stub** (real binary,
real store, real REST, real console, real metering — only the model is scripted).

1. **Full drag → brief → dispatch walk, in the browser.** Dragged a To-do card
   into Planning in the console; it went `planning → in_progress → in_review`,
   landing with a 2-step plan, its estimates, its verification and its scope on
   the card, and `[system] planned in 2 steps — everything it needs is in place,
   handing it to ceo` on the note.
2. **Missing prerequisite, readable on the board.** A pass claiming `slack` and
   an unconfigured MCP server returned the card to To-do, kept the brief, and
   the board itself renders both gaps with the fix for each ("connect it from
   the Connections tab" / "add it from the MCP tab").
3. **Usage.** 5 planning passes → 5 `planningCall` samples, all `agent=company`
   ($0.1155); 4 dispatch turns → 4 `inference` samples, all `agent=ceo`
   ($0.0924). The Usage view shows `company` and `Chief Executive` as separate
   rows — **zero planning samples under the assignee.** Finances posts both to
   the `Inference` category, memo'd `company` and `ceo`.
4. **`kill -9` mid-pass.** With the stub hanging, killed the host while the pass
   was in flight; after restart the card sits in To-do carrying *"the host
   restarted during planning, so the pass never finished — drag the card back
   into Planning to try again"*, with no plan.
5. **Secret safety.** Stored a real Composio token through the console route,
   re-ran a pass, and grepped the captured prompt: the token never appears —
   only `Composio credential configured: true`. The wire request carried no
   `tools` key at all.

6. **The mounted console, re-verified after the rebase.** All three headline
   states rendered on real cards: *blocked* (2 missing → red "Plan 2" tab,
   "needs 2 things" badge), *unresolved* (nothing missing but one prerequisite
   the host could not check → amber "Plan 1", "1 to be aware of"), and *all
   clear* (green, no count). The card carrying both features shows my badge and
   #339's link row together — `Planned · 2 steps ‖ Open CHANGELOG-v2.md` — and
   clicking #339's link still routes to `?artifact=…&v=1` and focuses the
   **Artifacts** tab, with the Plan tab sitting alongside it untouched.

Unit + integration: 29 tests in `src/harness/planning/test.rs`, covering every
verification arm, the parse (strict — prose is a failure, not a description),
the caps, the #339 interaction, and the whole pass against a real
`CompanyRuntime`.

The console has no unit-test harness in this repo (`frontend/package.json` has
no `test` script and no runner), so the three headline states are verified in a
real browser rather than by assertion — noted per the repo's
"document any intentionally untested edge case" rule.

## Impact

- **Risk accepted — auto-dispatch on success** spends the assignee's budget from
  a Planning drag. The drag is informed consent, the #304 per-agent cap still
  gates the dispatch turn, and the run still stops in In Review.
- **Risk accepted — a Composio/MCP outage yields `unknown`**, so a genuinely
  missing connection can slip through and the run fails with a real error. That
  is today's behaviour for every card, just rarer.
- **Prompt injection via card text is bounded, not eliminated**: typed parse,
  deterministic host-side verification, no tools, no secrets. The worst a
  hostile card achieves is a misleading brief a person reads.
- **A discarded pass costs one metered call** with no landing.
- **Estimates are model guesses, labelled as such.** Nothing derives budgets
  from them — the real caps are `budget_usd_daily` and the capability tier, both
  from live meter reads.
- Default build unaffected: no planner is wired, the column stays inert exactly
  as before, and the boot sweep returns anything left there.

## Related

Closes #337 (epic #183 §4). Builds on #408, which shipped the status→column
half, and lands alongside #339 — see *Union with #339* below.

### Union with #339 (rebased onto `27c4cee`)

Rebased onto current `main` after #339, #364, #374, #406 and #411 landed. The
whole conflict surface was **additive**: #339 added `TaskRecord.output` in the
same trailing slot this PR adds `TaskRecord.plan`, so all 30 hunks were "keep
both". Two struct literals then needed the other feature's field, and both are
`None` for a stated reason rather than as a default — #339's baseline fixture
exists to prove the output stamp round-trips against a card carrying nothing
else, and a card entering Planning has never run so it has produced nothing to
link to.

The interaction worth a test is the settle: #337 writes `plan`, #339 writes
`output`, and the planning settle read-modify-writes the live card rather than
rebuilding it, so `output` survives. Nothing in the code makes that coupling
visible — the settle never mentions `output` at all — and it is that silence
which has to keep being true, so
`a_re_plan_does_not_erase_what_an_earlier_attempt_produced` pins it. An
operator must not lose the link to finished work by asking for the card to be
re-planned.

### Review feedback

**CodeRabbit — "do not report unverified prerequisites as ready"** (comment
`3724614351`). Correct, and it mattered more than *minor*: the host is careful
to keep `unknown` out of `missing` precisely so a provider outage cannot
fabricate a blocker, and the headline then collapsed `unknown` back into
"everything is in place" — throwing that care away at the last step and
claiming a verification that never happened. `needsApproval` was the same
shape: the work *will* stop and ask, and the brief said it would not.

Now three states, on one shared `tallyPrerequisites` so the board badge, the
tab count and the headline cannot disagree:

| State | Headline | Badge | Tab |
|---|---|---|---|
| something missing | *"can't start yet — N things missing"* | *"needs N things"* | red count |
| nothing blocking, something unresolved | *"nothing is blocking it, but N couldn't be checked"* | *"N to be aware of"* | amber count |
| everything checked and passed | *"everything it needs was checked and is in place"* | *"Planned · N steps"* | no count |

The middle wording names which kind rather than lumping them — "couldn't be
checked" and "will stop for your approval" call for different responses from
the operator.

### Disagreements with the approved plan, and what shipped instead

1. **"Extract the connections projection" was already done.** #316 made
   `project_connections` a shared `pub(crate)` helper for the REST and GraphQL
   planes. The planner calls it rather than re-extracting, so there is still one
   definition of "connected". (It does make a live Composio probe with a 5s
   timeout; that is why an unreachable inventory yields `unknown`.)
2. **The planner is constructed in `runtime/builder.rs`, not `harness/build.rs`.**
   `build.rs` is about `build_agent`; the provider is minted in the builder's
   harness arm. `TaskPlanner::from_deps` is called there using the
   capture-out-of-the-arm pattern `wf_runner` / `steer_registry` /
   `run_supervisor` already use.
3. **`tokens_in` needed a one-line change** to make the plan's stated tier
   decision true — see *API Or Behavior Changes*.
4. **The `permission` verdict resolves a desk to its lead** rather than
   reporting `unknown`. The lead is who runs the turn, so the lead's grants are
   the ones that decide. A desk with no lead is still `unknown`.

## Summary by CodeRabbit

* **New Features**
  * Added a Planning stage for tasks, generating structured plans with steps, estimates, risks, verification criteria, and prerequisite checks.
  * Tasks with satisfied prerequisites can proceed automatically; blocked or failed planning returns them to To-do with clear explanations.
  * Added a read-only plan summary to task details.
  * Planning usage is tracked against inference budgets.

* **Bug Fixes**
  * Interrupted Planning tasks are recovered safely at startup and returned to To-do.
  * Added safeguards against duplicate planning and stale updates.

* **Documentation**
  * Documented Planning behavior, prerequisites, recovery, and task transitions.

### CODE REVIEW

File: frontend/src/views/TaskPlanBrief.tsx [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+        className={cn(
+          "mt-0.5 size-4 shrink-0",
+          blocked ? "text-destructive" : "text-emerald-600 dark:text-emerald-400",
+        )}
+      />
+      <div className="min-w-0 text-sm leading-relaxed">
+        <span className="font-medium">
+          {blocked
+            ? `Planned, but it can't start yet — ${blockers} thing${blockers === 1 ? "" : "s"} missing`
+            : "Planned, and everything it needs is in place"}
```
[the remark this answers is not in the document]
Comment (oxoxDev) [reply frontend/src/views/TaskPlanBrief.tsx]: Fixed in `5fe34d0`.

Agreed, and it mattered more than *minor*. The host is deliberately careful to keep `unknown` out of `missing` — an unreachable inventory must not fabricate a blocker, or a Composio outage would make every card in the company unplannable. The headline then collapsed `unknown` back into "everything is in place", which threw that care away at the last step and claimed a verification that never happened. `needsApproval` was the same shape: the work *will* stop and ask, and the brief said it would not.

Three states now, on one shared `tallyPrerequisites` so the board badge, the tab count and the headline cannot drift apart:

| State | Headline |
|---|---|
| something `missing` | *"can't start yet — N things missing"* (red) |
| nothing blocking, something unresolved | *"nothing is blocking it, but N couldn't be checked"* (amber) |
| everything checked and passed | *"everything it needs was checked and is in place"* (green) |

The middle wording names which kind rather than lumping them — "couldn't be checked" and "will stop for your approval" call for different responses from the operator. Verified all three in a browser against real cards.
